spec/awscr-s3/file_uploader_spec.cr

```diff
@@ -0,0 +1,89 @@
+require "../../spec_helper"
+
+module Awscr::S3
+  describe FileUploader do
+    describe "when the file is smaller than 5MB" do
+      it "uploads it in one call" do
+        WebMock.stub(:put, "http://s3.amazonaws.com/bucket/object?")
+               .with(body: "document")
+               .to_return(status: 200, body: "", headers: {"ETag" => "etag"})
+
+        client = Client.new("us-east-1", "key", "secret")
+        uploader = FileUploader.new(client)
+        small_io = IO::Memory.new("document")
+
+        uploader.upload("bucket", "object", small_io)
+      end
+
+      it "passes additional headers, when provided" do
+        WebMock.stub(:put, "http://s3.amazonaws.com/bucket/object?")
+               .with(body: "document", headers: {"x-amz-meta-name" => "myobject"})
+               .to_return(status: 200, body: "", headers: {"ETag" => "etag"})
+
+        client = Client.new("us-east-1", "key", "secret")
+        uploader = FileUploader.new(client)
+        small_io = IO::Memory.new("document")
+
+        uploader.upload("bucket", "object", small_io, {"x-amz-meta-name" => "myobject"})
+      end
+    end
+
+    describe "when the file is larger than 5MB" do
+      it "uploads it in chunks" do
+        # Start multipart upload
+        WebMock.stub(:post, "http://s3.amazonaws.com/bucket/object")
+               .with(query: {"uploads" => ""})
+               .to_return(status: 200, body: Helpers.start_multipart_upload_response(upload_id: "123"))
+
+        # Upload part 1
```

Comment:
the upload part stuff is a bit brittle because the `spawn` inside the `MultipartFileUploader` kind of swallows any webmock errors. removing this still makes the test fail, but only because of the expectation of the 2 parts in the *Complete multipart upload* stub below.

in the future, it would be nice to catch and re-raise erros in the spawn channel insde the `MultipartFileUploader` - knowing that something failed here is probably important for the user + it will also play nice with WebMock.

Reply:
So this test will pass every time assuming webmock is used? Or are you saying sometimes it fails because of spawn? Great idea about raising exceptions in the spawn, can be handled in another PR though imo

Reply:
if you break the part upload, the test would fail, but only because of the expectation in the last WebMock stub where I expect the 2 `<PartNumber>` items in the request body.

if we were to raise errors out of the `spawn` calls, WebMock would pinpoint exactly what happend. but yeah, that should be a different PR.